### PR TITLE
eth, les, geth: implement cli-configurable global gas cap for RPC calls

### PR DESCRIPTION
**Pull request by holiman**
This PR adds a CLI-configurable global gas cap on `eth_call` and `eth_estimateGas`. Although node operators should not expose rpc api to untrusted users (like, the internet), even with http level load balancing in front it is difficult to protect against dos via eth_call variants.
 
This makes it possible to enforce a cap on the gas that a caller can use, by using the flag `rpc.gascap`

## eth_call

Start geth
```
build/bin/geth --dev --rpc --rpc.gascap 33000
```
Caller:
```
[user@work go-ethereum]$ curl 127.0.0.1:8545 -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","id":1,"method":"eth_call","params":[{"data":"0x641fffffffe06000f3","gas":"0xffffffffffffffff"},"latest"]}'
{"jsonrpc":"2.0","id":1,"error":{"code":-32000,"message":"out of gas (gas capped at: 33000)"}}

```
Server: 
```
WARN [04-05|11:54:35.515] Applying cap on gas, caller requested amount above limit cap=33000 requested=0xffffffffffffffff
WARN [04-05|11:54:35.515] Served eth_call                          conn=127.0.0.1:44770 reqid=1 t=325.624µs err="out of gas (gas capped at: 33000)"
```

## eth_estimategas

caller
```
[user@work go-ethereum]$ curl  127.0.0.1:8545 -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","id":1,"method":"eth_estimateGas","params":[{"data":"0x641fffffffe06000f3","gas":"0xffffffffffffffff"}]}'
{"jsonrpc":"2.0","id":1,"error":{"code":-32000,"message":"gas required exceeds allowance (33000) or always failing transaction"}}
```
server
```
WARN [04-05|11:56:46.578] Applying cap on gas, caller requested amount above limit cap=33000 requested=0xffffffffffffffff
WARN [04-05|11:56:46.579] Served eth_estimateGas                   conn=127.0.0.1:44786 reqid=1 t=983.184µs err="gas required exceeds allowance (33000) or always failing transaction"
```

### CODE REVIEW

File: cmd/geth/main.go [hunk 1]
```diff
@@ -58,6 +58,7 @@ var (
 		utils.UnlockedAccountFlag,
 		utils.PasswordFileFlag,
 		utils.InsecureUnlockAllowedFlag,
+		utils.RPCGlobalGasCap,
```
Comment (karalabe) [inline cmd/geth/main.go]: Nitpick, but could we group the RPC flags together?

Comment (karalabe) [reply cmd/geth/main.go]: There's already an `rpcFlags` group, just a few leaked out. Wanna fix it? :)

File: node/config.go [hunk 2]
```diff
@@ -187,6 +188,9 @@ type Config struct {
 	staticNodesWarning     bool
 	trustedNodesWarning    bool
 	oldGethResourceWarning bool
+
+	// RPCGasCap is the global gas cap for eth-call variants.
+	RPCGasCap *big.Int `toml:",omitempty"`
```
Comment (karalabe) [inline node/config.go]: This should really not be a field on node.Config, rather on eth.Config.

File: node/service.go [hunk 3]
```diff
@@ -74,6 +75,12 @@ func (ctx *ServiceContext) ExtRPCEnabled() bool {
 	return ctx.config.ExtRPCEnabled()
 }
 
+// RPCGasCap returns the global cap for gas allowance on eth-call variants
+// over rpc. This can be used to protect against DoS attacks
+func (ctx *ServiceContext) RPCGasCap() *big.Int {
+	return ctx.config.RPCGasCap
+}
```
Comment (karalabe) [inline node/service.go]: Please move this out of `node`. The node is a dump thing that start arbitrary services with RPC APIs and networking layer. It should be oblivious to internals of what is actually starts.

Comment (karalabe) [reply node/service.go]: PS If you move the field into eth.Config, you won't even need this method at all.

**Comment by holiman** [comment]
Updated, ptal

File: internal/ethapi/api.go [hunk 4]
```diff
@@ -806,6 +820,13 @@ func DoEstimateGas(ctx context.Context, b Backend, args CallArgs, blockNr rpc.Bl
 // EstimateGas returns an estimate of the amount of gas needed to execute the
 // given transaction against the current pending block.
 func (s *PublicBlockChainAPI) EstimateGas(ctx context.Context, args CallArgs) (hexutil.Uint64, error) {
+	if gasCap := s.b.RPCGasCap(); gasCap != nil {
+		if new(big.Int).SetUint64(uint64(*args.Gas)).Cmp(gasCap) > 0 {
```
Comment (karalabe) [inline internal/ethapi/api.go]: args.Gas can be `nil`. This node will crash in that case. We handle gas illness inside `DoEstimateGas`. Perhaps we should pass the gascap as a parameter to `DoEstimateGas` and handle everything there.

File: internal/ethapi/api.go [hunk 5]
```diff
@@ -752,8 +752,22 @@ func DoCall(ctx context.Context, b Backend, args CallArgs, blockNr rpc.BlockNumb
 // Call executes the given transaction on the state for the given block number.
 // It doesn't make and changes in the state/blockchain and is useful to execute and retrieve values.
 func (s *PublicBlockChainAPI) Call(ctx context.Context, args CallArgs, blockNr rpc.BlockNumber) (hexutil.Bytes, error) {
-	result, _, _, err := DoCall(ctx, s.b, args, blockNr, vm.Config{}, 5*time.Second)
-	return (hexutil.Bytes)(result), err
+	if gasCap := s.b.RPCGasCap(); gasCap != nil {
+		if new(big.Int).SetUint64(uint64(*args.Gas)).Cmp(gasCap) > 0 {
```
Comment (karalabe) [inline internal/ethapi/api.go]: args.Gas can be `nil`. This node will crash in that case. We handle gas illness inside `DoEstimateGas`. Perhaps we should pass the gascap as a parameter to `DoEstimateGas` and handle everything there.

**Comment by holiman** [comment]
Thanks, and sorry about that! Fix implemented as you suggested. The error for estimategas is still helpful>
```
curl  127.0.0.1:8545 -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","id":1,"method":"eth_estimateGas","params":[{"data":"0x641fffffffe06000f3","gas":"0xffffffffffffffff"}]}'
{"jsonrpc":"2.0","id":1,"error":{"code":-32000,"message":"gas required exceeds allowance (50000) or always failing transaction"}}

```

However, the error code for `eth_Call` now does not tell the caller about the cap:
```
 curl 127.0.0.1:8545 -H "Content-Type: application/json" --data '{"jsonrpc":"2.0","id":1,"method":"eth_call","params":[{"data":"0x641fffffffe06000f3","gas":"0xffffffffffffffff"},"latest"]}'
{"jsonrpc":"2.0","id":1,"error":{"code":-32000,"message":"out of gas"}}

```
... which is less nice. Upside is that now it also fixed the problem via graphql too

**Pull request merged**